### DISCUSSION TITLE
General questions about this tool

### DISCUSSION THREAD

**Discussion by SQL-SQuirreL** [draivin/hsnips#180] [Q&A]
Just did some copy+pasting and added some JS logic to my snippet. And wow, it is really awesome what you can achieve with this extension in VS Code. Great work!
But I absolutely do not understand how to use syntax and how it fits all together, so I wondered if there is some tutorial that teaches about the basics. I even struggle on processing things I want to enter as parameters.

So here are my questions:

1. When I create a snippet, how can I achieve, that every character I enter for $1 - $n will be assigned to the resp. parameter, until I press TAB?
2. Is there a way, that I can make the things I've entered disappear in the editor window? E.g. I want to enter parameters $1 - $3. And after doing so, I want to show the processed string, I created via JavaScript only. The parameters I used to create the output string, which contains them as well, should not be shown in the text editor window.
3. Is there a way to show hints for the parameters, e.g. $1 being name, $2 being age etc. ?
4. How does the language specific settings work? Is there a connectiong between  LANG.hsnips prefix and file.LANG postfix?
Or does the extension only work for dedicated file types / languages?

If there is a guide for teaching the basics, I'd be happy about the link.

Thanks in advance and regards

**Comment by FluffLescure** [comment]
As for I know - which isn't very far - there is no tutorial, so most of it you'll have to learn by looking at other's snippets or finding out yourself. 

Usually snippets trigger around a keyword, so they don't involve parameters `$n` input parameters. What usually happens is after triggering that keyword, fields `$1` to `$n` are created to be written from. Here's an example:  
```
snippet for "for loop snippet" A
for $1 to $2, do $3
endsnippet
```
When the trigger word `for` is called you get the following seperate fields `for | to |, do |`, where the cursor hops from `$1` to `$3` .

That said, Hypersnips provides a return value `rv` and an array `m[ ]`  which contains information about the trigger that can be used inside the snippet using regex cases. 
Here's an example: 
```
snippet `for([0-9])([0-9])(.*)` "second for loop snippet" b
``rv=m[0]`` : for ``rv=m[1]`` to ``rv=m[2]``, do ``rv=m[3]``
endsnippet
```  
When the trigger word `for19something` is called after pressing TAB, this will yield `for19something : for 1 to 9, do something`. `m[0]` is the trigger itself and ``m[n]`` are each regex pattern found from the trigger.

I should also note that inside ` ``  `` ` you can create javascript code where `m[n]` are processed if you want to use them as arguments to a function. Here's an example I found for a matrix: 
```
snippet `M([1-9])x([1-9])` "matrix" A
``rv = '\\begin{pmatrix}'``
``count = 1;
mi = 0;
mj = 1;
msg = '';

for (i=0; i < parseInt(m[1],10); i++) {
  mi++;
  msg +=  mi;
  msg += mj;
  count++;
  for (j=1; j < parseInt(m[2],10); j++) {
    mj++;
    msg += '&\t' + mi + mj;
    count++;
  }
  mj = 1;
  if (i != parseInt(m[1],10)-1) {
    msg += ' \\\\\\ \n';
  }
}
rv = msg;``
\\end{pmatrix}
endsnippet
```

Showing hints about each separate `$n` field isn't supported in the sense that it is displayed in the VSC suggestion bar. However you can provide default values to each field which can help you. This is done following the sintax `${n:val}`. 

Going back to my first for snippet, if I replaced `$1` by `${1:start}`, this would produce `for start to |, do |`  where `start` would already be entirely selected and replaceable at the next keystroke. 

For the LANG.hsnips stuff, I'm not too sure yet how that works but I'm guessing there's something that matches the corresponding languagues to their respective file extension. you could check out https://github.com/draivin/hsnips/discussions/139 which touches upon it inside the all.hsnips file.
